SVN 1.7 Support, ignore svn:externals on local mod check and new switch toggle

bbaumer opened the pull request:
Fix: Calling svn info to determine if dest is an svn Working Copy, to support updates in Subfolders with Subversion > 1.7
Fix: Ignoring svn:externals on local Modification Check.
Add: Added Argument switch to alow skipping the svn switch call.

mpdehaan commented:
Can you share an example that replicates the problem that this addresses?

Thanks!

bbaumer commented:
With SVN 1.7 theres only one directory at the top-level of the Working Directory.
http://subversion.apache.org/docs/release-notes/1.7.html#single-db

### test-svn-1.7-support.yml

---
- hosts: localhost
  tasks:
  - name: Update SVN Repo
    subversion: repo=https://github.com/ansible/ansible-modules-core/trunk/source_control
                dest=~/repos/github/ansible-modules-core-test/ansible-modules-core/trunk/source_control
                force=no

### ansible-playbook test-svn-1.7-support.yml

PLAY [localhost] *************************************************************\* 

GATHERING FACTS **************************************************************\* 
ok: [localhost]

TASK: [Update SVN Repo] ******************************************************\* 
failed: [localhost] => {"failed": true}
msg: ERROR: /home/bbaumer/repos/github/ansible-modules-core-test/ansible-modules-core/trunk/source_control folder already exists, but its not a subversion repository.

FATAL: all hosts have already failed -- aborting

helium78 commented:
I tested this with Ansible 1.7.2 and svn 1.8.8 on Ubuntu 14.10 and get different results. Checking out works one time. Switching force=yes doesn't help

```
PLAY [localhost] ************************************************************** 

GATHERING FACTS *************************************************************** 
ok: [localhost]

TASK: [Update SVN Repo] ******************************************************* 
failed: [localhost] => {"cmd": "/usr/bin/svn --non-interactive --trust-server-cert --no-auth-cache switch https://github.com/ansible/ansible-modules-core/trunk/source_control /home/timo/repos/github/ansible-modules-core-test/ansible-modules-core/trunk/source_control", "failed": true, "rc": 1}
stderr: svn: E175002: Unexpected HTTP status 500 'Internal Server Error' on '/ansible/ansible-modules-core/!svn/bc/5244/trunk/source_control'

svn: E175002: Additional errors:
svn: E175002: REPORT request on '/ansible/ansible-modules-core/!svn/bc/5244/trunk/source_control' failed: 500 Internal Server Error

msg: svn: E175002: Unexpected HTTP status 500 'Internal Server Error' on '/ansible/ansible-modules-core/!svn/bc/5244/trunk/source_control'

svn: E175002: Additional errors:
svn: E175002: REPORT request on '/ansible/ansible-modules-core/!svn/bc/5244/trunk/source_control' failed: 500 Internal Server Error

FATAL: all hosts have already failed -- aborting

PLAY RECAP ******************************************************************** 
           to retry, use: --limit @/home/timo/noext.retry

localhost                  : ok=1    changed=0    unreachable=0    failed=1   

```

Disabling the externals should be optional. I need them to be checked out ;-)

bbaumer commented:
The Problem with svn switch seems to be something with the Github Infrastructure. The SVN Client gets an  HTTP 500 on the HTTP REPORT Request. 
svn switch is called after the Repos has been checked out on the first Playbook run. 

Against the Subversion Source Repo the svn switch works.

`subversion: repo=http://svn.apache.org/repos/asf/subversion/trunk
                dest=~/repos/github/subversion
                force=no
`

force=yes is default and only applies if you have changes in your Working Copy.

svn:externals are only ignored, when the Modul tries to detect changes in your Working Copy, not on checkout or update.

bbaumer commented:
Is there something else I can do to help integrating this pull request?

mpdehaan commented:
Hi @bbaumer, we have a lot of pull requests in the queue and this was just submitted a week ago.

It may take a while to get into a release, but we appreciate it.

Thanks!

gregdek commented:
@bbaumer Thanks for submitting this pull request. Please accept our apologies for taking so long to review it.

Unfortunately, due to this delay, it appears that the underlying code has changed, and this PR needs to be rebased. If you believe that this PR still applies, please rebase and we will then put it into our (new, faster) community review process as soon as possible.  Again, apologies for the delay and the extra work.

bbaumer committed:
Fix: Calling svn info to determine if dest is an svn Working Copy,
to support updates in Subfolders with Subversion > 1.8
Fix: Ignoring svn:externals on local Modification Check.
Add: Added Argument switch to alow skipping the svn switch call.

bbaumer commented:
@gregdek Yes, I believe that this PullRequest still applies. I have rebased the PR today.
It would be great, if I will see my contribution in upstream soon.

I made few tests with the new switch Parameter and force (yes and no) after the Rebase. All my Test are successfull.

gregdek commented:
Thanks @bbaumer. 

@dsummersl please review; if you give +1, we will promote to priority_reviewed.

dsummersl commented:
will do.

dsummersl commented on source_control/subversion.py:
```diff
@@ -115,8 +122,20 @@ def _exec(self, args):
+            return out.splitlines()
+        else:
+            rc, out, err = self.module.run_command(bits, check_rc)
+            return rc
```
To make this a little more DRY, move the line `rc, out, err = self.module.run_command(bits, check_rc)` in front of the if/else block (so that there is only one instance of this command).

dsummersl commented on source_control/subversion.py:
```diff
@@ -115,8 +122,20 @@ def _exec(self, args):
+        if rc == 0:
+            return True
+        else:
+            return False
```
This can simplify to `return rc == 0`.

dsummersl commented on source_control/subversion.py:
```diff
@@ -103,7 +110,7 @@ def __init__(
 
-    def _exec(self, args):
+    def _exec(self, args, check_rc=True):
         bits = [
```
Document the changes to this function - having a parameter change what the function returns could be confusing. Something along the lines of:

```
Execute a subversion command, and return output. If check_rc is False, returns the return code instead of the output.
```

dsummersl commented:
@bbaumer looks pretty good - have a few small suggestions from CR.

dsummersl commented on source_control/subversion.py:
```diff
@@ -78,6 +78,13 @@
+    required: false
+    default: "yes"
+    choices: [ "yes", "no" ]
+    version_added: "1.6"
```
Probably v1.9 now?

bcoca replied:
2.0

bbaumer replied:
Should I remove this new function and pull a version without it first?

bbaumer committed:
version_added for switch Parameter changed to 2.0
Add comment to explain check_rc Parameter in _exec Function
Optimize code and clean up is_svn_repo Function

dsummersl commented:
@gregdek +1

gregdek commented:
Ah, missed this +1 earlier. Sorry for the delay. Marking for inclusion. Thanks @dsummersl for the review and @bbaumer for the PR.

gregdek commented:
(ping @ansible/core)

jimi-c commented:
Merged, thanks!
